linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,180 @@
+
+- increment the flag signaling that the threads can complete.
+
+- close the epoll object.
+
+- close the thread by calling `ThreadAPI_Join`
+
+- then the memory associated with `completion_port` shall be freed.
+
+### completion_port_inc_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_inc_ref` handles the incrementing of the reference count of the module
+
+If `completion_port` is `NULL`, `completion_port_inc_ref` shall return.
+
+Otherwise `completion_port_inc_ref` shall increment the internally maintained reference count.
+
+### completion_port_post_queue
+
+```C
+MOCKABLE_FUNCTION(, int, completion_port_post_queue, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
+```
+
+`completion_port_post_queue` adds an ingress item to the post queue for epoll signaling
+
+If `completion_port` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+If `ingress_data` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+`completion_port_post_queue` shall allocate a thread data object
+
+`completion_port_post_queue` shall allocate a thread data object
+
+`completion_port_post_queue` shall add the socket in the epoll system by calling `epoll_ctl` with `EPOLL_CTL_MOD` along with the `COMPLETION_PORT_INGRESS_DATA` `epoll_op` variable.
+
+If the `epoll_ctl` call fails with `ENOENT`, `completion_port_post_queue` shall call `epoll_ctl` again with `EPOLL_CTL_ADD`.
+
+On success, `completion_port_post_queue` shall return 0.
+
+If any error occurs, `completion_port_post_queue` shall fail and return a non-zero value.
+
+### completion_port_dequeue
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dequeue, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+If `completion_port` is `NULL`, `completion_port_dequeue` shall return.
+
+If `socket` is `INVALID_SOCKET`, `completion_port_dequeue` shall return.
+
+`completion_port_dequeue` shall remove the underlying socket from the epoll by calling  `epoll_ctl` with `EPOLL_CTL_DEL`.
+
+### epoll_worker_func
+
+```c
+static int epoll_worker_func(void* parameter)
+```
+
+If `parameter` is `NULL`, `epoll_worker_func` shall do nothing.
+
+`epoll_worker_func` shall call `epoll_wait` to wait for an epoll event to become signaled.
+
+If the events value contains `EPOLLRDHUP` (hang up), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then call the `event_callback` callback with the `event_callback_ctx` and `COMPLETION_PORT_RESULT_ABANDONED`.
+
+If the events value contains `EPOLLIN` (recv), `epoll_worker_func` shall the following:
+
+- `epoll_worker_func` shall receive the `EPOLL_THREAD_DATA` value from the ptr variable from the `epoll_event` data ptr.
+
+- Then `epoll_worker_func` shall call `recv` and do the following:
```

Comment:
>recv [](http://example.com/codeflow?start=39&length=4)

I see this module doing the `recv` and `send` - how's that different from ASYNC_SOCKET which I remember having reviewd a recent while ago...

Reply:
ASYNC_SOCKET will use this module so it won't have to do recv and sends anymore.